project/api.py

```diff
@@ -0,0 +1,204 @@
+    code = 404
+    return jsonify(error=str(e), code=code)
+
+
+@app.errorhandler(400)
+def handle_bad_request(e):
+    """Handle cases with wrong query parameter"""
+    code = 400
+    return jsonify(error=str(e), code=code), code
+
+
+@app.route('/api/v1/queries/avg_elapsed_time', methods=['GET'])
+def avg_elapsed_time():
+    """Returns an average elapsed time for all queries on certain date"""
+    query_parameters = request.args
+    date = query_parameters.get('date')
+    if not date:
+        return page_not_found('No date provided')
+
+    data = pd.read_csv('queries.tsv', sep='\t')
+    try:
+        date_start = datetime.strptime(date, '%Y-%m-%d')
+    except ValueError as e:
+        return handle_bad_request(e)
+    date_end = date_start + timedelta(days=1, milliseconds=-1)
+    date_start_code = to_milliseconds(date_start)
+    date_end_code = to_milliseconds(date_end)
+
+    day_data = data[data['time'].between(date_start_code, date_end_code)]
+    if day_data.shape[0] == 0:
+        return handle_bad_request('Date is out of range')
+
+    day_pivot = day_data.pivot(index='query_id', columns='status', values='time')
+    not_closed_within_day = day_pivot.isnull().values.any(1).nonzero()[0]  # if not empty returns indexes
+
+    if not_closed_within_day.shape[0] > 0:
+        day_pivot.drop(day_pivot.index[not_closed_within_day], inplace=True)
+
+    day_pivot['elapsed_time'] = (day_pivot[1] - day_pivot[0]) / 1000
+
+    avg_time = day_pivot['elapsed_time'].mean()
+
+    return jsonify({'Average elapsed time, sec': avg_time})
+
+
+@app.route('/api/v1/queries/avg_rows_per_sec', methods=['GET'])
+def avg_rows_per_sec():
+    """Returns an average per second number of rows returned from all queries during certain time range"""
+    query_parameters = request.args
+    time_range_start = query_parameters.get('start')
+    time_range_end = query_parameters.get('end')
+    if not time_range_start or not time_range_end:
+        return page_not_found('No time range provided')
+
+    try:
+        datetime_start_code, datetime_end_code = prepare_request(time_range_start, time_range_end)
+    except ValueError as e:
+        return handle_bad_request(e)
+
+    data = pd.read_csv('queries.tsv', sep='\t')
+
+    time_range_data = data[data['time'].between(datetime_start_code, datetime_end_code)]
+    if time_range_data.shape[0] == 0:
+        return handle_bad_request(400)
+
+    time_range_pivot = time_range_data.pivot(index='query_id', columns='status', values=['time', 'rows'])
+
+    not_closed_within_range = time_range_pivot.isnull().values.any(1).nonzero()[0]  # if not empty returns indexes
+
+    if not_closed_within_range.shape[0] > 0:
+        time_range_pivot.drop(time_range_pivot.index[not_closed_within_range], inplace=True)
+
+    time_range_pivot['elapsed_time'] = (time_range_pivot['time'][1] - time_range_pivot['time'][0]) / 1000
+
+    avg_rows = time_range_pivot['rows'][1].sum() / time_range_pivot['elapsed_time'].sum()
```

Comment:
Here the `elapsed_time` for different queries may overlap. You should take that into account if you want to get average per second **from all queries**.

Reply:
So if, for example, query 1 was executed between seconds 2 and 6 and query 2 was executed between seconds 4 and 8, their total elapsed time will be 8-2 and average will be (8-2)/2?

Reply:
You need to average the rows here. For the case
```
second 2 -> 10 rows from query 1
second 3 -> 10 rows from query 1
second 4 -> 10 rows from query 1, 20 rows from query 2
second 5 -> 10 rows from query 1, 20 rows from query 2
second 6 -> 10 rows from query 1, 20 rows from query 2
second 7 -> 20 rows from query 2
second 8 -> 20 rows from query 2
```
the result will be
```
(10 + 10 + 30 + 30 + 30 + 20 + 20) / 8
```